File: part 5 - operators precedence/practice/Interpreter.py
from Lexer import Lexer
from Token import Token


class Interpreter(object):
    def __init__(self, text):
        self.lexer = Lexer(text)
        self.current_token = self.lexer.get_next_token()

    def error(self):
        raise Exception('Error parsing input')

    def compare(self, token_types):
        if self.current_token.type in token_types:
            self.current_token = self.lexer.get_next_token()
        else:
            self.error()

    def factor(self):
        """<factor> ::= <number> | "(" <expression> ")" """
        if self.current_token.type == Token.INTEGER:
            result = self.current_token.value
            self.compare([Token.INTEGER])
        elif self.current_token.type == Token.LEFTPARENTHESIS:
            self.compare([Token.LEFTPARENTHESIS])
            result = self.expr()
            self.compare([Token.RIGHTPARENTHESIS])
        else:
            result = None
            self.error()

        return result

    def term(self):
        """<term> ::= <factor> { ("*" | "/") <factor> }*"""
        result = self.factor()

        while self.current_token.type in [Token.MULTIPLY, Token.DIVIDE]:
            op = self.current_token
            self.compare([Token.MULTIPLY, Token.DIVIDE])

            right = self.factor()

            result = self.eval(result, op, right)

        return result

    def eval(self, left, op, right):
        """<eval> ::= <term> <op> <term>"""

        if op.type == Token.MULTIPLY:
            return left * right
        elif op.type == Token.DIVIDE:
            return left / right
        elif op.type == Token.PLUS:
            return left + right
        elif op.type == Token.MINUS:
            return left - right

    def expr(self):
        """<expression> ::= <term> { ("+" | "-") <term> }*"""

        result = self.term()

        while self.current_token.type in [Token.PLUS, Token.MINUS]:
            op = self.current_token
            self.compare([Token.PLUS, Token.MINUS])

            right = self.term()

            result = self.eval(result, op, right)

        return int(result)

File: part 5 - operators precedence/practice/Interpreter.cs
using System;
using test.@base;

namespace test.part_5.practice
{
    public class Interpreter : IInterpreter
    {
        Lexer lexer { get; init; }

        public Interpreter(string text)
        {
            lexer = new Lexer(text);
            lexer.Tokenize();
            lexer.Position = 0;
        }

        private static void Error(string text)
        {
            throw new Exception($"Invalid element exception. {text} was not expected.");
        }

        private int Factor()
        {
            if (lexer.Tokens[lexer.Position].Type == Token.INT)
            {
                return int.Parse(lexer.Tokens[lexer.Position++].Value);
            } else if (lexer.Tokens[lexer.Position++].Type == Token.LPAREN)
            {
                int result = Eval();
                lexer.Position++;
                return result;
            } else
            {
                Error(lexer.Tokens[lexer.Position].Value);
                return -1;
            }
        }

        private int Term()
        {
            int result = Factor();

            while (lexer.Position < lexer.Tokens.Count()
                && (lexer.Tokens[lexer.Position].Type == Token.MULTIPLY || lexer.Tokens[lexer.Position].Type == Token.DIVIDE))
            {
                string op = lexer.Tokens[lexer.Position++].Value;
                int right = Factor();

                result = Eval(result, op, right);
            }

            return result;
        }

        private static int Eval(int left, string op, int right)
        {
            switch(op)
            {
                case "+":
                    return left + right;

                case "-":
                    return left - right;

                case "*":
                    return left * right;

                case "/":
                    return left / right;

                default:
                    Error(op);
                    return -1;
            }
        }

        public int Eval()
        {
            int result = Term();

            while (lexer.Position < lexer.Tokens.Count() &&
                (lexer.Tokens[lexer.Position].Type == Token.PLUS || lexer.Tokens[lexer.Position].Type == Token.MINUS))
            {
                string op = lexer.Tokens[lexer.Position++].Value;
                int right = Term();
                result = Eval(result, op, right);
            }

            return result;
        }
    }
}
